sql: implement sequence caching and cached sequence serial normalization

jayshrivastava opened the pull request:
### sql: implement sequence caching

Previously, incrementing sequences at a high throughput
would result in many distributed writes to the KV layer
due to MVCC. This has caused garbage collection problems
in the past. This would occur in situations such as
bulk importing data while using the sequence number as an
id for each new row being added.

This change allows clients to cache sequence numbers in their local
session data. When the cache is empty, the sequence will be
incremented once by the cache size * increment amount, which are
both sequence options. Then, all the intermediate values will be
cached locally on a node to be given out whenever the sequence is
incremented.

To accommodate schema changes, cached sequences values will be
invalidated when new descriptor versions are seen by the cache.
This invalidation can occur when old versions are seen as well
to accommodate schema change rollbacks.

Release note (sql change): Using the CACHE sequence option no longer
results in an "unimplemented" error. The CACHE option is now fully
implemented and will allow nodes to cache sequence numbers. A cache
size of 1 means that there is no cache, and cache sizes of less than 1
are not valid.

### sql: create benchmark for concurrent sequence increments

Previously, there was no benchmark that tested the performance
of concurrent increments to sequences. There was also no
benchmark which compared sequence performance based on
different cache sizes. This change updates a benchmark to measure
performance based on the above criteria.

Release note: None

### sql: add serial normalization setting for cached sequences

Closes: #51259

Release note (sql change): The `serial_normalization` session
variable can now be set to the value `sql_sequence_cached`.
Cached sequences will allow nodes to cache 256 sequence numbers
locally. The underlying sequence will only be incremened once (by
256 increments) when the cache is empty. Using `sql_sequence_cached`
will result in better performance than `sql_sequence` because the
former will perform fewer distributed calls to increment sequences.
However, cached seqences may contribute to large gaps between
sequence numbers if cached values are lost due to errors or
node outages.

###  sql: make cached sequences bounds-related semantics match pg semantics

Previously, cached sequences did not obey pg semantics
with respect to exceeding bounds. For example, if there
were a sequence with a cache size of 5 and max value of 2,
calling nextval(...) would immediately cause an error due
to exceeded bounds. According to postgres, nextval(...)
should return 1, then 2, then finally return an error due
to the max value of 2 being reached. This change alters
sequence caching behavior to match the above semantics.

Additionally, this change now makes it possible for sequences
to exceed the bounds set by MAXVALUE and MINVALUE. This is
because calling nextval(...) should be as fast as possible,
and the fastest way to do this is to let nextval(...) always
succeed on incrementing a sequence. Despite this, a user
will never see any values that exceed a sequence's bounds.
To make a sequence incrementable again after exceeding its
bounds, there are two options:
1. The user changes the sequence's value by calling setval(...).
2. The user performs a schema change to alter the sequences MinValue
   or MaxValue. If the value of a sequence exceeds its bounds,
   it must be restored to the original MinValue or MaxValue in
   the same transaction as the schema change.
This change adds code to handle case 2 above.

Release note: None

### Benchmark
Using `make bench PKG='./pkg/sql' BENCHES='BenchmarkSequenceIncrement' TESTFLAGS="--count=5 --benchmem" |& tee c{size} .txt`:

Caching 256 values
```
name                     old time/op    new time/op    delta
SequenceIncrement/P-1-8     313µs ± 1%     107µs ± 1%  -65.85%  (p=0.008 n=5+5)
SequenceIncrement/P-2-8     327µs ± 2%      82µs ± 2%  -75.04%  (p=0.008 n=5+5)
SequenceIncrement/P-4-8     347µs ± 1%      70µs ± 2%  -79.89%  (p=0.008 n=5+5)
SequenceIncrement/P-8-8     389µs ± 2%      65µs ± 3%  -83.40%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
SequenceIncrement/P-1-8     130kB ± 1%      57kB ± 1%  -55.76%  (p=0.008 n=5+5)
SequenceIncrement/P-2-8     131kB ± 0%      49kB ± 1%  -62.56%  (p=0.008 n=5+5)
SequenceIncrement/P-4-8     132kB ± 0%      46kB ± 0%  -65.26%  (p=0.008 n=5+5)
SequenceIncrement/P-8-8     134kB ± 0%      44kB ± 1%  -67.09%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
SequenceIncrement/P-1-8       807 ± 0%       406 ± 0%  -49.75%  (p=0.000 n=5+4)
SequenceIncrement/P-2-8       812 ± 1%       363 ± 0%  -55.32%  (p=0.008 n=5+5)
SequenceIncrement/P-4-8       826 ± 1%       346 ± 0%  -58.10%  (p=0.008 n=5+5)
SequenceIncrement/P-8-8       863 ± 1%       341 ± 0%  -60.44%  (p=0.008 n=5+5)
```

For the other cache sizes I tested, the performance improvement is virtually the same as above. 
[c1.txt](https://github.com/cockroachdb/cockroach/files/5692305/c1.txt)
[c32.txt](https://github.com/cockroachdb/cockroach/files/5692306/c32.txt)
[c64.txt](https://github.com/cockroachdb/cockroach/files/5692307/c64.txt)
[c128.txt](https://github.com/cockroachdb/cockroach/files/5692308/c128.txt)
[c256.txt](https://github.com/cockroachdb/cockroach/files/5692310/c256.txt)
[c512.txt](https://github.com/cockroachdb/cockroach/files/5692326/c512.txt)
[c1024.txt](https://github.com/cockroachdb/cockroach/files/5692309/c1024.txt)

jayshrivastava committed:
sql: implement sequence caching

Previously, incrementing sequences at a high throughput
would result in many distributed writes to the KV layer
due to MVCC. This has caused garbage collection problems
in the past. This would occur in situations such as
bulk importing data while using the sequence number as an
id for each new row being added.

This change allows clients to cache sequence numbers in their local
session data. When the cache is empty, the sequence will be
incremented once by the cache size * increment amount, which are
both sequence options. Then, all the intermediate values will be
cached locally on a node to be given out whenever the sequence is
incremented.

To accommodate schema changes, cached sequences values will be
invalidated when new descriptor versions are seen by the cache.
This invalidation can occur when old versions are seen as well
to accommodate schema change rollbacks.

Release note (sql change): Using the CACHE sequence option no longer
results in an "unimplemented" error. The CACHE option is now fully
implemented and will allow nodes to cache sequence numbers. A cache
size of 1 means that there is no cache, and cache sizes of less than 1
are not valid.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56954)
<!-- Reviewable:end -->

jayshrivastava committed:
sql: add serial normalization setting for cached sequences

Closes: https://github.com/cockroachdb/cockroach/issues/51259

Release note (sql change): The `serial_normalization` session
variable can now be set to the value `sql_sequence_cached`.
If this value is set, the cluster setting
`sql.defaults.serial_sequences_cache_size` can be used to
control the number of values to cache in a user's session with a
default of 256. When the cache is empty, the the underlying
sequence will only be incremened once to populate it. Using
`sql_sequence_cached` will result in better performance than
`sql_sequence` because the former will perform fewer distributed
calls to increment sequences. However, cached seqences may
result in large gaps between serial sequence numbers if a session
terminates before using all the values in its cache.

ajwerner reviewed (commented):
> Reviewed 2 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/sql/exec_util.go, line 745 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MNPPfnk1J9N4qPySzZY:-MNPPfnl8tU9CPA86r4J:bl7c1n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eba15800a7c27ffd5b1ee73f18a88ad365bbf6/pkg/sql/exec_util.go#L745)):*
> ```Go
> 	VersionUpgradeHook func(ctx context.Context, to roachpb.Version) error
> 
> 	SequenceCache *SequenceCache
> ```

give this a comment to indicate that it is node-level cache of sequence values.
___
*[pkg/sql/sequence.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MNPOXhZEvriRumGxLMH:-MNPOXhZEvriRumGxLMI:b-80247y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eba15800a7c27ffd5b1ee73f18a88ad365bbf6/pkg/sql/sequence.go#L43)):*
> ```Go
> // are stored using sequence descriptor IDs as keys. The cache should only
> // be accessed using the provided API.
> type SequenceCache struct {
> ```

let's iterate on this a bit. I think it's worthwhile to write some microbenchmarks to help guide the optimizations here.

I think it's a bad idea to hold the mutex during the entire operation for a given cache value. Instead let's utilize a singleflight.Group for initializing and updating the cache value. 

Another thing we will likely eventually want to do is to switch to a `syncutil.IntMap` but let's leave that off for the first optimization path. 
___
*[pkg/sql/sequence.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MNPP55b44SZYHuov3ie:-MNPP55b44SZYHuov3if:boss5dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eba15800a7c27ffd5b1ee73f18a88ad365bbf6/pkg/sql/sequence.go#L64)):*
> ```Go
> 		currentIndex int
> 		// Stores values to be given out.
> 		values *[]int64
> ```

rather than using a slice, I think we can get away with constant memory usage. What if we had a curVal, increment, maxVal or something like that. 
___
*[pkg/sql/sequence.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MNPQEan3SD3TyFteryx:-MNPQEan3SD3TyFteryy:b9p9gjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eba15800a7c27ffd5b1ee73f18a88ad365bbf6/pkg/sql/sequence.go#L186)):*
> ```Go
> 		}
> 
> 		if seqOpts.CacheSize == 1 {
> ```

all previously existing sequences will have a value of `0`. How do we deal with that? Can you write a test that injects such a descriptor and ensure that it behaves properly?


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
sql: create benchmark for concurrent sequence increments

Previously, there was no benchmark that tested the performance
of concurrent increments to sequences. There was also no
benchmark which compared sequence performance based on
different cache sizes. This change adds a benchmark to measure
performance based on the above criteria.

Release note: None

ajwerner reviewed (changes requested):
Seems like you need a rebase. Sorry I took so darn long to get back to this. It's very close. 

> Reviewed 1 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MSe4uhu8RyNFkSMgG-m:b90a8tt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/sql/exec_util.go, line 783 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSe555D9n_xsK9p62Kt:-MSe555D9n_xsK9p62Ku:bvy1g2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/exec_util.go#L783)):*
> ```Go
> 
> 	// VersionUpgradeHook is called after validating a `SET CLUSTER SETTING
> 	// version` but before executing it. It structuredcan carry out arbitrary migrations
> ```

detritus?
___
*[pkg/sql/exec_util.go, line 2250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSe58ZkBxmHiQY7_jBE:-MSe58ZkBxmHiQY7_jBF:b-m95pcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/exec_util.go#L2250)):*
> ```Go
> }
> 
> // initSequenceCache
> ```

Missing a comment.
___
*[pkg/sql/sequence.go, line 105 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSeGtC29yWKuKxvRasd:-MSeGtC29yWKuKxvRase:b-mhghdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/sequence.go#L105)):*
> ```Go
> 		val = int64(rowid)
> 	} else {
> 		fetchNextValues := func() (int64, int64, int64, error) {
> ```

Consider pulling everything in this else into a helper. 
___
*[pkg/sql/sequence.go, line 224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSeH9TZ1Uxl6Q47pLAH:-MSeH9TZ1Uxl6Q47pLAI:be5xlh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/sequence.go#L224)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	// If there are cached values, changing the value of the sequence should invalidate the cache.
> 	// For simplicity, the cache invalidates implicitly when it sees new descriptor versions. Thus,
> 	// a schema change is triggered here to make sure the cache gets invalidated.
> 	mutableDescriptor, err := resolver.ResolveMutableExistingTableObject(ctx, p, seqName, true, tree.ResolveRequireSequenceDesc)
> ```
</details>

I don't think we need to do this. Simpler not to. Sorry I sent you astray here. 

>ALTER SEQUENCE will not immediately affect nextval results in backends, other than the current one, that have preallocated (cached) sequence values. They will use up all cached values prior to noticing the changed sequence generation parameters. The current backend will be affected immediately.

https://www.postgresql.org/docs/12/sql-altersequence.html


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/sql/exec_util.go, line 783 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSe555D9n_xsK9p62Kt:-MSiZbk_EIHKeZPmAs8g:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/exec_util.go#L783)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

detritus?
</blockquote></details>

Yes.
___
*[pkg/sql/exec_util.go, line 2250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSe58ZkBxmHiQY7_jBE:-MSi_3vL6BsgZpyrc1Hw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/exec_util.go#L2250)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Missing a comment.
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSeH9TZ1Uxl6Q47pLAH:-MSiZpvQ9U_v2n6YOTtA:b-ltkh9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/110cb700f856ab2c1891c3f5d618dd4c3c1bd131/pkg/sql/sequence.go#L224)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// If there are cached values, changing the value of the sequence should invalidate the cache.
> 	// For simplicity, the cache invalidates implicitly when it sees new descriptor versions. Thus,
> 	// a schema change is triggered here to make sure the cache gets invalidated.
> 	mutableDescriptor, err := resolver.ResolveMutableExistingTableObject(ctx, p, seqName, true, tree.ResolveRequireSequenceDesc)
> ```
I don't think we need to do this. Simpler not to. Sorry I sent you astray here. 

>ALTER SEQUENCE will not immediately affect nextval results in backends, other than the current one, that have preallocated (cached) sequence values. They will use up all cached values prior to noticing the changed sequence generation parameters. The current backend will be affected immediately.

https://www.postgresql.org/docs/12/sql-altersequence.html
</blockquote></details>

Yup it's simpler not to.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
This is great! @arulajmani is going to take a review pass too. Really just final touches now. Seems like it still needs a fresh rebase. 

> Reviewed 4 of 12 files at r2, 4 of 11 files at r6, 1 of 7 files at r9, 1 of 5 files at r11, 3 of 7 files at r12, 1 of 1 files at r13, 5 of 5 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MSkPbZx1dMffekWmp5t:b-ckyyo2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[pkg/sql/sequence.go, line 124 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSkOkk83VkmXnX3TARD:-MSkOkk83VkmXnX3TARE:bxtkmfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/24b7752117e8aa678b80e276679cda0a6a43140b/pkg/sql/sequence.go#L124)):*
> ```Go
> 	}
> 
> 	fetchNextValues := func() (int64, int64, int64, error) {
> ```

nit: name the return values. 
___
*[pkg/sql/sequence_test.go, line 405 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSkPFWJ0wD456J5etbD:-MSkPFWJ0wD456J5etbE:b-5dl16i) ([raw file](https://github.com/cockroachdb/cockroach/blob/24b7752117e8aa678b80e276679cda0a6a43140b/pkg/sql/sequence_test.go#L405)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 	execStmt := func(statement string) {
> 		sqlSessions[anySession()].Exec(t, statement)
> 	}
> 	checkIntValue := func(session int, statement string, value int) {
> 		sqlSessions[session].CheckQueryResults(t, statement, [][]string{
> 			{fmt.Sprintf("%d", value)},
> 		})
> 	}
> ```
</details>

plumb `t` into these explicitly, otherwise you're going to be using the `t` from the parent test rather than the subtest which is not allowed. 
___
*[pkg/sql/serial.go, line 45 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MSkPTbu2Z4gaDkv00hR:-MSkPTbv0qRS2EpUo6Ih:b-mphbqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/24b7752117e8aa678b80e276679cda0a6a43140b/pkg/sql/serial.go#L45)):*
> ```Go
> // cachedSequencesCacheSize is the default cache size used when
> // SessionNormalizationMode is SerialUsesCachedSQLSequences.
> var cachedSequencesCacheSize int64 = 256
> ```

I could see making this a cluster setting. 


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Sounds good! I keep having to rebase because of changes to the descriptor protobuf haha.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MSnvDPt36TBGwb6T67C:bucjwfj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (changes requested):
Exciting stuff! :100: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MT1PDp47tMAACVAv0HT:b-uqyrno)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/sql/sequence.go, line 129 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1M7hl1xg1teB85qKT:-MT1M7hl1xg1teB85qKU:b7p7dc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L129)):*
> ```Go
> }
> 
> // incrementSequenceUsingCache fetches the next value of the sequence represented by the passed catalog.TableDescriptor.
> ```

nit: wrap at 80 chars
___
*[pkg/sql/sequence.go, line 141 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1MI6u9rhKxgkSNOk2:-MT1MI6u9rhKxgkSNOk3:b-v60w4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L141)):*
> ```Go
> 	// If a sequence has a cache size of 0, it should be treated in the same was as sequences
> 	// with cache sizes of 1.
> 	cacheSize := seqOpts.CacheSize
> ```

Can we move this into a function (maybe in `structured.go`?) and do the "0 means 1" translation there, as opposed to special casing here?
___
*[pkg/sql/sequence.go, line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1NCBWAw09SllaCIUs:-MT1NCBWAw09SllaCIUt:b-y31ort) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L158)):*
> ```Go
> 			return 0, 0, 0, err
> 		}
> 		if endValue > seqOpts.MaxValue || endValue < seqOpts.MinValue {
> ```

I'm curious how this works in PG -- does it error out if the last cache value exceeds bounds, or does it truncate the cache? 

This does deserve some tests though. 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1399 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1KDyJ9Pu9BgJGNlhM:-MT1KDyJ9Pu9BgJGNlhN:b-xx262d) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/logictest/testdata/logic_test/sequences#L1399)):*
> ```
> # Verify cache invalidation with schema changes.
> 
> # 10 values (1,2,...,10) are cached, and the underlying sequence is increment to 10.
> ```

nit: incremented?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1403 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1P8jDDx9yUM4rPMrO:-MT1P8jDDx9yUM4rPMrP:b-5fq9yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/logictest/testdata/logic_test/sequences#L1403)):*
> ```
> SELECT nextval('cache_test')
> ----
> 1
> ```

nit: might be worth sanity checking `curval` and `lastval` interactions here as well


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/sequence.go, line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1NCBWAw09SllaCIUs:-MT2qLQE4BgSlgMzOCBj:b6y9h8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L158)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I'm curious how this works in PG -- does it error out if the last cache value exceeds bounds, or does it truncate the cache? 

This does deserve some tests though. 
</blockquote></details>

Say there's a sequence with MAXVALUE 50 and CACHE 100. In PG, the first time the cache gets populated, the underlying sequence is incremented to 50 and only 50 values are cached. If a new session calls nextval(...) again, an error occurs because the underlying sequence is at the MAXVALUE. 

I believe implementing this means that we would have to change kv.IncrementValRetryable to use the planner transaction to first read the underlying sequence, check for errors, then write a new value. This adds more contention on sequence increments, but thats okay because we added caching as a method to avoid contention. 

I can implement the changes above and write a test. I just wanted to confer with you first :) Let me know if what I said sounds good. Thanks for taking a look at this!


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/sequence.go, line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1NCBWAw09SllaCIUs:-MT69Dss5O_IJsrJ7K_M:b-59awmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L158)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Say there's a sequence with MAXVALUE 50 and CACHE 100. In PG, the first time the cache gets populated, the underlying sequence is incremented to 50 and only 50 values are cached. If a new session calls nextval(...) again, an error occurs because the underlying sequence is at the MAXVALUE. 

I believe implementing this means that we would have to change kv.IncrementValRetryable to use the planner transaction to first read the underlying sequence, check for errors, then write a new value. This adds more contention on sequence increments, but thats okay because we added caching as a method to avoid contention. 

I can implement the changes above and write a test. I just wanted to confer with you first :) Let me know if what I said sounds good. Thanks for taking a look at this!
</blockquote></details>

Why would we need to use the planner transaction? It feels like you can continue to deal with passing the maxvalue here and then if you increase `MaxValue` then you could reset it to the old maxvalue transactionally. 


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/sequence.go, line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1NCBWAw09SllaCIUs:-MT6QoDoEQvQHaDVRnnZ:bav0cbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L158)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Why would we need to use the planner transaction? It feels like you can continue to deal with passing the maxvalue here and then if you increase `MaxValue` then you could reset it to the old maxvalue transactionally. 
</blockquote></details>

From my understanding, `p.Txn` is what we use to perform a transactional reads/writes. What would be the alternatives?

By "increase `MaxValue`", do you mean exceed `MaxValue`? My approach was to read the sequence value first, but this reactive approach also works and is probably better since we avoid the read every time we increment. Just to be clear, are you suggesting something similar to the following?
```
		endValue, err := kv.IncrementValRetryable(
			ctx, p.txn.DB(), seqValueKey, seqOpts.Increment*cacheSize)

		if err != nil {
			if errors.HasType(err, (*roachpb.IntegerOverflowError)(nil)) {
				return 0, 0, 0, boundsExceededError(descriptor)
			}
			return 0, 0, 0, err
		}
		if endValue > seqOpts.MaxValue || endValue < seqOpts.MinValue {
			// This sequence was already at the MaxValue before incrementing,
			// so it must be reset transactionally.
			if endValue - seqOpts.Increment*cacheSize == seqOpts.MaxValue {
				boundsError := boundsExceededError(descriptor)
				err = p.txn.Put(ctx, seqValueKey, seqOpts.MaxValue)
				if err != nil {
					boundsError = errors.CombineErrors(boundsError, err)
				}
				return 0, 0, 0, boundsError
			}

			// The sequence was incremented over the MaxValue, so it must be
			// set to the MaxValue transactionally. The cache must store values
			// up to and including the MaxValue.
			err = p.txn.Put(ctx, seqValueKey, seqOpts.MaxValue)
			if err != nil {
				return 0, 0, 0, err
			}
			return seqOpts.MaxValue, seqOpts.Increment, (seqOpts.Increment*cacheSize - (endValue - seqOpts.MaxValue)) / seqOpts.Increment, nil
```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/sequence.go, line 158 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MT1NCBWAw09SllaCIUs:-MT72V4L0fQp2iW28bUl:bm8xsjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/34720d62864b46609cfde5c18d1fe16e0a54825e/pkg/sql/sequence.go#L158)):*
> 		// The sequence was incremented over the MaxValue, so it must be
> 			// set to the MaxValue transactionally. The cache must store values
> 			// up to and including the MaxValue.
> 			err = p.txn.Put(ctx, seqValueKey, seqOpts.MaxValue)
> 			if err != nil {
> 				return 0, 0, 0, err
> 			}

I wouldn't even do this here. Just let the value keep getting run up once it has exceeded the maximum value. Then, in `ALTER SEQUENCE ... SET MAXVALUE` if the value is going to be increased, you could do a transactional read-modify-write to see if it is larger than the current `MAXVALUE` and then reset it to the old `MAXVALUE`. Does that make sense?

This way subsequent attempts to increment will also keep increasing the value but returning an error until they overflow but that will also just return an error. One thing is that `currval` should return `MAXVALUE` if the sequence is above `MAXVALUE`.


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
sql: make cached sequences bounds-related semantics match pg semantics

Previously, cached sequences did not obey pg semantics
with respect to exceeding bounds. For example, if there
were a sequence with a cache size of 5 and max value of 2,
calling nextval(...) would immediately cause an error due
to exceeded bounds. According to postgres, nextval(...)
should return 1, then 2, then finally return an error due
to the max value of 2 being reached. This change alters
sequence caching behavior to match the above semantics.

Additionally, this change now makes it possible for sequences
to exceed the bounds set by MAXVALUE and MINVALUE. This is
because calling nextval(...) should be as fast as possible,
and the fastest way to do this is to let nextval(...) always
succeed on incrementing a sequence. Despite this, a user
will never see any values that exceed a sequence's bounds.
To make a sequence incrementable again after exceeding its
bounds, there are two options:
1. The user changes the sequence's value by calling setval(...).
2. The user performs a schema change to alter the sequences MinValue
   or MaxValue. If the value of a sequence exceeds its bounds,
   it must be restored to the original MinValue or MaxValue in
   the same transaction as the schema change.
This change adds code to handle case 2 above.

Release note: None

ajwerner reviewed (commented):
My one remaining thing on this is do we need a cluster version to gate this in the mixed version state. This week is the last week before stability period. I don't have much doubt that this is going to make it across the line. Thanks for all the revisions!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MTcYim8AyJD39AtmmDG:b-swzkc4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @jayshrivastava)

___
*[pkg/sql/sequence.go, line 171 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTcYHJxENBymU6Z4Zqr:-MTcYHJxENBymU6Z4Zqs:b-jdldzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b7f02b250072255987a0224d2c978f972dce6f/pkg/sql/sequence.go#L171)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 
> 			if endValue < seqOpts.MinValue {
> 				if endValue-seqOpts.Increment*cacheSize <= seqOpts.MinValue {
> 					return 0, 0, 0, boundsExceededError(descriptor)
> ```
</details>

What is this case dealing with?
___
*[pkg/sql/sequence.go, line 176 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTcYLky9zDhlLmYH9rc:-MTcYLky9zDhlLmYH9rd:b95tmvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b7f02b250072255987a0224d2c978f972dce6f/pkg/sql/sequence.go#L176)):*
> ```Go
> 					return 0, 0, 0, boundsExceededError(descriptor)
> 				}
> 				return endValue - seqOpts.Increment*(cacheSize-1), seqOpts.Increment, int64(math.Abs(float64(seqOpts.MinValue-(endValue-seqOpts.Increment*cacheSize)))) / seqOpts.Increment * -1, nil
> ```

It feels like there's some repetition that can be dealt with here. 


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Having a cluster in a mixed version state is fine. 

Caching occurs in a session, so the feature is isolated from other nodes. The new code handles all cache sizes, including 0. The old code will ignore a sequence's cache size when incrementing a sequence. In the case of creating/altering a sequence's cache size, the old code will just throw an error saying that caching is not supported. 

One tricky thing is that a session running the new code may increment a sequence beyond its MAXVALUE several times. A user talking to another node may try to increase the MAXVALUE so that the sequence can be incremented again. If the other node is running the new version, any value greater than the old max value will make the sequence incrementable again. This is because the new alter sequence code will restore the value of the sequence to the old MAXVALUE before executing the schema change. If the node has an old version, this will not happen. To make the sequence incrementable again, a user will have to set a max value greater than the current value of the sequence or they will have to use setval() to set the value of the sequence to the old MAXVALUE first. Overall, I think this is okay since sequences are non-transactional and we are allowed to skip numbers.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MTgdNALCPklLNxfzSv7:bs4d5lj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/sequence.go, line 171 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTcYHJxENBymU6Z4Zqr:-MTgbZYg1lbeYVu6DL5w:bf1cv3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b7f02b250072255987a0224d2c978f972dce6f/pkg/sql/sequence.go#L171)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 			if endValue < seqOpts.MinValue {
> 				if endValue-seqOpts.Increment*cacheSize <= seqOpts.MinValue {
> 					return 0, 0, 0, boundsExceededError(descriptor)
> ```
What is this case dealing with?
</blockquote></details>

Just repetition of the above case when seqOpts.Increment < 0.  I just updated the code to not have this repetition. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Get rid of the float `math.Abs` stuff and this is ready for bors!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MTi9Qpo3AhByh3SNAez:b-6uyrbw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @jayshrivastava)

___
*[pkg/sql/sequence.go, line 148 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTi8RxR05QfDdrEqX4F:-MTi8RxR05QfDdrEqX4G:bgmvq4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/192bac796e23cbc7e0e14eff7bc4e631b56f6495/pkg/sql/sequence.go#L148)):*
> ```Go
> 	cacheSize := seqOpts.EffectiveCacheSize()
> 
> 	fetchNextValues := func() (currentValue int64, incrementAmount int64, sizeOfCache int64, err error) {
> ```

nit, you can group the return values: `(currentValue, incrementAmount, sizeOfCache int64, err error)
___
*[pkg/sql/sequence.go, line 164 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTi7x_55xf0s2hJiD18:-MTi7x_55xf0s2hJiD19:b8sr84y) ([raw file](https://github.com/cockroachdb/cockroach/blob/192bac796e23cbc7e0e14eff7bc4e631b56f6495/pkg/sql/sequence.go#L164)):*
> ```Go
> 		if endValue > seqOpts.MaxValue || endValue < seqOpts.MinValue {
> 			// If the sequence exceeded its bounds prior to the increment, then return an error.
> 			if seqOpts.Increment > 0 && endValue-seqOpts.Increment*cacheSize >= seqOpts.MaxValue || seqOpts.Increment < 0 && endValue-seqOpts.Increment*cacheSize <= seqOpts.MinValue {
> ```

nit: wrap this conditional. 
___
*[pkg/sql/sequence.go, line 172 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-MTi7o6eDytnlvtE3Q30:-MTi7o6eDytnlvtE3Q31:b2mg7vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/192bac796e23cbc7e0e14eff7bc4e631b56f6495/pkg/sql/sequence.go#L172)):*
> ```Go
> 				limit = seqOpts.MinValue
> 			}
> 			return endValue - seqOpts.Increment*(cacheSize-1), seqOpts.Increment, int64(math.Abs(float64(limit-(endValue-seqOpts.Increment*cacheSize))) / math.Abs(float64(seqOpts.Increment))), nil
> ```

Also, please assign these variables for readability. 

this math.Abs stuff seems like it'll get screwed up as these limit values get to be larger than can be represented in a float64. Just write an `
```
abs := func(i int64) int64 {
    if i < 0 { 
        return -i
    } 
    return i
}
```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks again for keeping after this. This is going to be a great addition!

bors r+

> Reviewed 3 of 13 files at r20, 4 of 5 files at r22, 1 of 3 files at r23, 1 of 1 files at r25.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56954#-:-MTkQVH8C8mn1PyFXBsp:b-hbf1d0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)



<!-- Sent from Reviewable.io -->

mosquito2333 commented on pkg/sql/sequence.go:
```diff
@@ -116,28 +116,88 @@ func incrementSequenceHelper(
-		if val > seqOpts.MaxValue || val < seqOpts.MinValue {
-			return 0, boundsExceededError(descriptor)
+
+		// This sequence has exceeded its bounds after performing this increment.
```
Why should out-of-range endValue be stored in the cache?

sandr8 commented:
> Cached sequences will allow nodes to cache 256 sequence numbers locally

This is not the behavior I have observed. At least on v21.2.10, the sequence seems to be cached per-session (as compared to per-node). This means that -unless sessions are very long lived- a lot of identifiers will be wasted and not a lot of improvement in terms of contention will be achieved.

rafiss commented:
You're right - this cache is implemented per-session. The [sequence docs](https://www.cockroachlabs.com/docs/stable/create-sequence.html#parameters) are a little more clear, but I agree that this release note is misleading, and the docs could have more clarity.

One thing to note is that this was implemented to be compatible with sequencee caching in PostgreSQL, which is described here in the "Notes" section: https://www.postgresql.org/docs/current/sql-createsequence.html

And I also agree with you that the sequence cache size should be chosen based on how frequently the sequence is used along with how long the session lasts.

ale-dd commented:
@rafiss I agree. But I would love the addition of a keyword to enable the per-node caching, which would be very advantageous.

Some syntax like `[ [ PER NODE ] CACHE # ]`...

rafiss commented:
I've filed https://github.com/cockroachdb/cockroach/issues/89338 to track that request.
